File: src/org/jetbrains/jet/j2k/ast/CallChainExpression.java
package org.jetbrains.jet.j2k.ast;

import org.jetbrains.annotations.NotNull;

/**
 * @author ignatov
 */
public class CallChainExpression extends Expression {
  private final Expression myExpression;
  private final Expression myIdentifier;

  public Expression getIdentifier() {
    return myIdentifier;
  }

  @NotNull
  @Override
  public Kind getKind() {
    return Kind.CALL_CHAIN;
  }

  public CallChainExpression(Expression expression, Expression identifier) {
    myExpression = expression;
    myIdentifier = identifier;
  }

  @Override
  public boolean isNullable() {
    return myIdentifier.isNullable();
  }

  @NotNull
  @Override
  public String toKotlin() {
    if (!myExpression.isEmpty())
      if (myExpression.isNullable())
        return myExpression.toKotlin() + QUESTDOT + myIdentifier.toKotlin();
      else
        return myExpression.toKotlin() + DOT + myIdentifier.toKotlin();
    return myIdentifier.toKotlin();
  }
}

File: src/org/jetbrains/jet/j2k/ast/CallChainExpression.kt
package org.jetbrains.jet.j2k.ast

public open class CallChainExpression(val expression : Expression, val identifier : Expression) : Expression() {
    public override fun isNullable() : Boolean {
        if (!expression.isEmpty() && expression.isNullable()) return true
        return identifier.isNullable()
    }

    public override fun toKotlin() : String {
        if (!expression.isEmpty()) {
            return expression.toKotlin() + (if (expression.isNullable()) "?." else ".") + identifier.toKotlin()
        }

        return identifier.toKotlin()
    }
}
